### ISSUE TITLE
HMR failed when using setState

### ISSUE THREAD

**Issue by ochervak** [developit/preact-boilerplate#81]
For minimal working code that reproduce issue:
```
import { h, Component } from 'preact';

export default class Home extends Component {
  state = { tick: 0 };

  componentDidMount() {
    setInterval(() => this.setState({ tick: this.state.tick + 1 }), 1000);
  }

  render({}, { tick }) {
    return (
	<div>hello {tick}</div>
    );
  }
}
```

Steps to reproduce:
 - Try to change layout in render()

Actual result:
Screen flickers twice, first time with old data (like previous state), than with new state

Expected:
Correct HMR replace of DOM.

![preact-boilerplate-bug](https://cloud.githubusercontent.com/assets/13982649/20464228/d2c90d22-af4c-11e6-8de8-4af807cfd648.gif)

<details>
 <summary>package.json</summary>

```javascript
{
  "name": "penguins-team-webapp",
  "version": "2.0.0",
  "description": "Progressive web application for Penguins Team",
  "scripts": {
    "dev": "cross-env NODE_ENV=development webpack-dev-server --inline --hot --progress",
    "start": "superstatic build -p 8080 --host 0.0.0.0 --gzip -c '{\"rewrites\": [{\"source\":\"**\",\"destination\":\"index.html\"}],\"headers\":[{\"source\":\"**\",\"headers\":[{\"key\":\"Cache-Control\",\"value\":\"max-age=86400\"}]}]}'",
    "prestart": "npm run build",
    "build": "cross-env NODE_ENV=production webpack -p --progress",
    "prebuild": "mkdirp build && ncp src/assets build/assets",
    "lint": "eslint {src,test}"
  },
  "keywords": [
    "preact"
  ],
  "license": "MIT",
  "author": "Ostap Chervak <ostap.chervak@strikersoft.com>",
  "devDependencies": {
    "autoprefixer": "^6.5.3",
    "babel": "^6.5.2",
    "babel-core": "^6.18.2",
    "babel-eslint": "^7.1.1",
    "babel-loader": "^6.2.8",
    "babel-plugin-transform-decorators-legacy": "^1.3.4",
    "babel-plugin-transform-react-jsx": "^6.8.0",
    "babel-preset-es2015": "^6.18.0",
    "babel-preset-es2015-minimal": "^2.1.0",
    "babel-preset-stage-0": "^6.16.0",
    "babel-register": "^6.18.0",
    "babel-runtime": "^6.18.0",
    "cross-env": "^3.1.3",
    "css-loader": "^0.26.0",
    "eslint": "^3.10.2",
    "extract-text-webpack-plugin": "^1.0.1",
    "file-loader": "^0.9.0",
    "html-webpack-plugin": "^2.24.1",
    "json-loader": "^0.5.4",
    "less": "^2.7.1",
    "less-loader": "^2.2.3",
    "mkdirp": "^0.5.1",
    "ncp": "^2.0.0",
    "node-sass": "^3.13.0",
    "path": "^0.12.7",
    "postcss-loader": "^1.1.1",
    "raw-loader": "^0.5.1",
    "sass-loader": "^4.0.2",
    "source-map-loader": "^0.1.5",
    "style-loader": "^0.13.1",
    "superstatic": "^4.0.3",
    "url-loader": "^0.5.7",
    "webpack": "^1.13.3",
    "webpack-dev-server": "^1.16.2"
  },
  "dependencies": {
    "preact": "^6.4.0",
    "preact-compat": "^3.9.2",
    "preact-router": "^2.3.2"
  }
}

 ```

</details>
<details>
 <summary>webpack.config.babel.js</summary>

```javascript
import webpack from 'webpack';
import ExtractTextPlugin from 'extract-text-webpack-plugin';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import autoprefixer from 'autoprefixer';
import path from 'path';

const ENV = process.env.NODE_ENV || 'development';

const CSS_MAPS = ENV!=='production';

module.exports = {
  context: path.resolve(__dirname, "src"),
  entry: './index.js',

  output: {
    path: path.resolve(__dirname, "build"),
    publicPath: '/',
    filename: 'bundle.js'
  },

  resolve: {
    extensions: ['', '.jsx', '.js', '.json', '.less'],
    modulesDirectories: [
      path.resolve(__dirname, "src/lib"),
      path.resolve(__dirname, "node_modules"),
      'node_modules'
    ],
    alias: {
      components: path.resolve(__dirname, "src/components"),		// used for tests
      style: path.resolve(__dirname, "src/style"),
      'react': 'preact-compat',
      'react-dom': 'preact-compat'
    }
  },

  module: {
    preLoaders: [
      {
        test: /\.jsx?$/,
        exclude: /src\//,
        loader: 'source-map'
      }
    ],
    loaders: [
      {
        test: /\.jsx?$/,
        exclude: /node_modules/,
        loader: 'babel'
      },
      {
        test: /\.(less|css)$/,
        include: /src\/components\//,
        loader: ExtractTextPlugin.extract('style?singleton', [
          `css?sourceMap=${CSS_MAPS}&modules&importLoaders=1&localIdentName=[local]${process.env.CSS_MODULES_IDENT || '_[hash:base64:5]'}`,
          'postcss',
          `less?sourceMap=${CSS_MAPS}`
        ].join('!'))
      },
      {
        test: /\.(less|css)$/,
        exclude: /src\/components\//,
        loader: ExtractTextPlugin.extract('style?singleton', [
          `css?sourceMap=${CSS_MAPS}`,
          `postcss`,
          `less?sourceMap=${CSS_MAPS}`
        ].join('!'))
      },
      {
        test: /\.(scss|css)$/,
        include: /src\/components\//,
        loader: ExtractTextPlugin.extract('style?singleton', [
          `css?sourceMap=${CSS_MAPS}&modules&importLoaders=1&localIdentName=[local]${process.env.CSS_MODULES_IDENT || '_[hash:base64:5]'}`,
          'postcss',
          `sass?sourceMap=${CSS_MAPS}`
        ].join('!'))
      },
      {
        test: /\.(scss|css)$/,
        exclude: /src\/components\//,
        loader: ExtractTextPlugin.extract('style?singleton', [
          `css?sourceMap=${CSS_MAPS}`,
          `postcss`,
          `sass?sourceMap=${CSS_MAPS}`
        ].join('!'))
      },
      {
        test: /\.json$/,
        loader: 'json'
      },
      {
        test: /\.(xml|html|txt|md)$/,
        loader: 'raw'
      },
      {
        test: /\.(svg|woff2?|ttf|eot|jpe?g|png|gif)(\?.*)?$/i,
        loader: ENV==='production' ? 'file?name=[path][name]_[hash:base64:5].[ext]' : 'url'
      }
    ]
  },

  postcss: () => [
    autoprefixer({ browsers: 'last 2 versions' })
  ],

  plugins: ([
    new webpack.NoErrorsPlugin(),
    new ExtractTextPlugin('style.css', {
      allChunks: true,
      disable: ENV!=='production'
    }),
    new webpack.optimize.DedupePlugin(),
    new webpack.DefinePlugin({
      'process.env': JSON.stringify({ NODE_ENV: ENV })
    }),
    new HtmlWebpackPlugin({
      template: './index.html',
      minify: { collapseWhitespace: true }
    })
  ]).concat(ENV==='production' ? [
    new webpack.optimize.OccurenceOrderPlugin()
  ] : []),

  stats: { colors: true },

  node: {
    global: true,
    process: false,
    Buffer: false,
    __filename: false,
    __dirname: false,
    setImmediate: false
  },

  devtool: ENV==='production' ? 'source-map' : 'cheap-module-eval-source-map',

  devServer: {
    port: process.env.PORT || 8080,
    host: '0.0.0.0',
    colors: true,
    publicPath: '/',
    contentBase: './src',
    historyApiFallback: true
  }
};

 ```

</details>

**Comment by developit** [comment]
Hey - can you test this with 7.0.3? `npm i -S preact@beta`

**Comment by ochervak** [comment]
@developit same problem with preact 7.0.3. Maybe i also need to install beta preact-compat ?

**Comment by developit** [comment]
Nah there isn't one. Any chance you have the devtools integration imported?

**Comment by ochervak** [comment]
@developit nope. just clean install of boilerplate and updated dependencies (listed in package.json) and support for sass (listed in webpack config)

**Comment by developit** [comment]
Interesting. This is the first I've seen of this issue, will look into it.

**Comment by developit** [comment]
Ah - you aren't cancelling your timer. Need to do that in `componentWillUnmount`. Your component is getting recycled, but the timer is still active in a zombie instance of that component and so still updates the UI.

**Comment by ochervak** [comment]
@developit just tried to do, as you mentioned, but it didn't work for me :(

```
componentDidMount() {
    this.timer = setInterval(() => this.setState({ tick: this.state.tick + 1 }), 1000);
  }

  componentWillUnmount() {
    clearInterval(this.timer);
  }
```

**Comment by developit** [comment]
I think this is fixed as of 7.1.0, but I'll be sure and check before closing the issue out :)

**Comment by developit** [comment]
Confirmed, this is fixed:

![preview](https://i.gyazo.com/aa63c2d056ce1ec5a35c412c58b8336b.gif)
